src/helpers/clipboard/clipboard_figure_state.ts

```diff
@@ -84,12 +84,20 @@ export class ClipboardFigureState implements ClipboardState {
    */
   paste(target: Zone[]) {
     const sheetId = this.getters.getActiveSheetId();
+    const size = { height: this.copiedFigure.height, width: this.copiedFigure.width };
+    const viewportRect = this.getters.getMainViewportRect();
+    const colDimensions = this.getters.getColDimensions(sheetId, target[0].left);
+    const rowDimensions = this.getters.getRowDimensions(sheetId, target[0].top);
     const position = {
-      x: this.getters.getColDimensions(sheetId, target[0].left).start,
-      y: this.getters.getRowDimensions(sheetId, target[0].top).start,
+      x:
+        viewportRect.width - colDimensions.start < size.width
+          ? viewportRect.width - size.width - 96
+          : colDimensions.start,
+      y:
+        viewportRect.height - rowDimensions.start < size.height
+          ? viewportRect.height - size.height - 28
+          : rowDimensions.start,
```

Comment:
Doesn't work wiht frozen pane. If there are frozen pane getMainViewportRect !== grid dimension. Ise the position of the last column/row instead.

Reply:
Doing somethink like `mainViewport.x + mainViewport.width` would probably work yeah, but you really shouldn't care here about viewports and whatnot, all you want is the end of the sheet. `getColDimension(lastCol)` is cleaner IMO.

Reply:
@jash-odoo   see here...